commit 2c0331f4f7d241995452b99afaf0aab00493334a
Author: Michael Contreras <michael@inetric.com>
Date:   Wed Dec 5 13:31:30 2012 -0500
Bug-Debian: http://bugs.debian.org/696051
Comment: second half of the fix for CVE-2012-6075
Comment: see also e1000-discard-packets-that-are-too-long-if-not-SBP-and-not-LPE.patch

    e1000: Discard oversized packets based on SBP|LPE
    
    Discard packets longer than 16384 when !SBP to match the hardware behavior.
    
    Signed-off-by: Michael Contreras <michael@inetric.com>
    Signed-off-by: Stefan Hajnoczi <stefanha@redhat.com>

diff --git a/hw/e1000.c b/hw/e1000.c
index 92fb00a..8fd1654 100644
--- a/hw/e1000.c
+++ b/hw/e1000.c
@@ -61,6 +61,8 @@ static int debugflags = DBGBIT(TXERR) | DBGBIT(GENERAL);
 
 /* this is the size past which hardware will drop packets when setting LPE=0 */
 #define MAXIMUM_ETHERNET_VLAN_SIZE 1522
+/* this is the size past which hardware will drop packets when setting LPE=1 */
+#define MAXIMUM_ETHERNET_LPE_SIZE 16384
 
 /*
  * HW models:
@@ -809,8 +811,9 @@ e1000_receive(NetClientState *nc, const uint8_t *buf, size_t size)
     }
 
     /* Discard oversized packets if !LPE and !SBP. */
-    if (size > MAXIMUM_ETHERNET_VLAN_SIZE
-        && !(s->mac_reg[RCTL] & E1000_RCTL_LPE)
+    if ((size > MAXIMUM_ETHERNET_LPE_SIZE ||
+        (size > MAXIMUM_ETHERNET_VLAN_SIZE
+        && !(s->mac_reg[RCTL] & E1000_RCTL_LPE)))
         && !(s->mac_reg[RCTL] & E1000_RCTL_SBP)) {
         return size;
     }
